torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -175,22 +141,74 @@ bool query(rank_type rank, const std::vector<int>& sockets) {
   }
 }
 
-bool checkAndUpdate(const store_type& store,
-                           std::vector<std::string>& keys) {
+bool Store::StoreDeamon::checkAndUpdate(std::vector<std::string>& keys) const {
   bool ret = true;
-  for (auto it = keys.begin(); it != keys.end(); it++) {
-    if (store.find(*it) == store.end())
+  for (auto it = keys.begin(); it != keys.end();) {
+    if (_store.find(*it) == _store.end()) {
       ret = false;
-    else
+      it++;
+    } else {
       it = keys.erase(it);
+    }
   }
   return ret;
 }
 
-void wake_up(int socket) {
-  send_value<QueryType>(socket, QueryType::STOP_WAITING);
+
+
+Store::Store(rank_type rank, const std::string& addr,
+             port_type port, rank_type world_size)
+ : _rank(rank)
+ , _store_addr(addr)
+ , _store_port(port)
+ , _socket(-1)
+ , _store_thread(nullptr)
+{
+  // only one process starts store
+  if (_rank == 0) {
+    _store_thread = std::unique_ptr<StoreDeamon>(
+      new StoreDeamon(port, world_size)
+    );
+  }
+
+  _socket = connect(_store_addr, _store_port);
 }
 
-} // anonymous namespace
+Store::~Store() {
+  // the 0 process has to stop the daemon
+  if (_rank == 0) {
+    send_value<QueryType>(_socket, QueryType::FINISH);
```

Comment:
Is this query needed anymore? It should be enough to disconnect from the store to make it exit